core/field_dropdown.js

```diff
@@ -68,15 +68,23 @@ Blockly.FieldDropdown = function(menuGenerator, opt_validator) {
   this.menuGenerator_ = menuGenerator;
 
   /**
-   * The currently selected index. A value of -1 indicates no option
-   * has been selected.
+   * A cache of the most recently generated options.
+   * @type {Array.<!Array.<string>>}
+   * @private
+   */
+  this.generatedOptions_ = null;
```

Comment:
Did you look into having this be a map/dictionary? I think there are some loops that could be elliminated that way.

Reply:
What would the key of this map be? If it's the value, it means you can't have multiple options with the same value, label, is the same. I don't think we set that restrictions right now.

Reply:
> What would the key of this map be?

It would be keyed by value. This is what the dropdown currently uses, it's just inefficient and uses for-loops instead of a map.

We may not currently throw errors, but the language-neutral keys of your dropdown have to be unique, otherwise your dropdown breaks:

![SameValues](https://user-images.githubusercontent.com/25440652/65188205-b8761380-da22-11e9-9bd0-1b119361404e.gif)

<details><summary>Steps to reproduce</summary>
<p>

1. Add the following block definition to the playground.
```
{
    "type": "test_dropdowns_same_value",
    "message0": "same value %1",
    "args0": [
      {
        "type": "field_dropdown",
        "name": "FIELDNAME",
        "options": [
          [ "first item", "ITEM1" ],
          [ "second item", "ITEM2" ],
          [ "third item", "ITEM2" ],
          [ "fourth item", "ITEM2" ],
          [ "fifth item", "ITEM2" ],
        ]
      }
    ],
    "style": "fields_dropdowns"
  }
```
2. Add the block to the workspace.
3. Use the dropdown.
4. Observe the bad behavior.
</p>
</details>

If we were to switch it to a map we could definitely get rid of the loop in `doClassValidation_`. We could also probably get rid of the `selectedIndex_` property and just use the value instead, elliminating the loop in `doValueUpdate_`.